### PR TITLE
BP-36: Stats documentation annotation

### PR DESCRIPTION
**Pull request by sijie**
Descriptions of the changes in this PR:

*Motivation*

A common ask from people using bookkeeper is how they can monitor bookies and bookkeeper clients, what kind of metrics that bookkeeper exposes
and what are the important metrics. Currently bookkeeper doesn't provide any metrics page for guiding people on monitoring bookkeeper services.

In order to help people on this, we need to provide a few documentation pages about metrics. However if we just write such pages, those pages
can quickly get out-of-dated when code is changed. The proposal here is to seek a programming way for generating metrics related pages.

Master Issue: #1785

> ---
> In order to uphold a high standard for quality for code contributions, Apache BookKeeper runs various precommit
> checks for pull requests. A pull request can only be merged when it passes precommit checks. However running all
> the precommit checks can take a long time, some trivial changes don't need to run all the precommit checks. You
> can check following list to skip the tests that don't need to run for your pull request. Leave them unchecked if
> you are not sure, committers will help you:
>
> - [ ] [skip bookkeeper-server bookie tests]: skip testing `org.apache.bookkeeper.bookie` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server client tests]: skip testing `org.apache.bookkeeper.client` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server replication tests]: skip testing `org.apache.bookkeeper.replication` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server tls tests]: skip testing `org.apache.bookkeeper.tls` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server remaining tests]: skip testing all other tests in bookkeeper-server module.
> - [ ] [skip integration tests]: skip docker based integration tests. if you make java code changes, you shouldn't skip integration tests.
> - [ ] [skip build java8]: skip build on java8. *ONLY* skip this when *ONLY* changing files under documentation under `site`.
> - [ ] [skip build java9]: skip build on java9. *ONLY* skip this when *ONLY* changing files under documentation under `site`.
> ---

> ---
> Be sure to do all of the following to help us incorporate your contribution
> quickly and easily:
>
> If this PR is a BookKeeper Proposal (BP):
>
> - [ ] Make sure the PR title is formatted like:
>     `<BP-#>: Description of bookkeeper proposal`
>     `e.g. BP-1: 64 bits ledger is support`
> - [ ] Attach the master issue link in the description of this PR.
> - [ ] Attach the google doc link if the BP is written in Google Doc.
>
> Otherwise:
> 
> - [ ] Make sure the PR title is formatted like:
>     `<Issue #>: Description of pull request`
>     `e.g. Issue 123: Description ...`
> - [ ] Make sure tests pass via `mvn clean apache-rat:check install spotbugs:check`.
> - [ ] Replace `<Issue #>` in the title with the actual Issue number.
> 
> ---

### CODE REVIEW

**Comment by sijie** [comment]
run bookkeeper-server replication tests

File: site/bps/BP-36-stats-documentation-annotation.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+### Public Interfaces
+
+Introduced a `StatsDoc` annotation.
+
+```bash
+/**
+ * Documenting the stats.
+ */
+@Retention(RetentionPolicy.RUNTIME)
```
Comment (eolivelli) [inline site/bps/BP-36-stats-documentation-annotation.md]: Why runtime?

**Comment by dlg99** [comment]
@sijie i.e. addEntry does bookieAddEntry and journalAddEntry so it is a parent for both.
ate the same time we want to show that bookieAddEntry happens before journalAddEntry to i.e. order these metrics correctly in the documentation and describe the sequence fo events there.

parent defines hierarchy and happensAfter simply orders sequence of siblings.

**Comment by sijie** [comment]
@dlg99 gotcha. I will update the BP

**Comment by sijie** [comment]
@dlg99 updated the BP with your suggestion. I will add those two in the annotation after #1787 is in.

File: site/bps/BP-36-stats-documentation-annotation.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    /**
+     * The metric name of an operation that happens
+     * after the operation of this metric.
+     *
+     * <p>similar as {@link #parent()}, it can be used for analyzing
+     * the dependencies between metrics.
+     *
+     * @return the metric name of an operation that happens after the operation of this metric.
+     */
+    default String happenAfter() { return ""; }
```
[the remark this answers is not in the document]
Comment (sijie) [reply site/bps/BP-36-stats-documentation-annotation.md]: ah, I was planning to type `happensAfter`. however it becomes `happenAfter` :(

[the remark this answers is not in the document]
Comment (sijie) [reply site/bps/BP-36-stats-documentation-annotation.md]: fixed

**Comment by sijie** [comment]
run pr validation

**Comment by sijie** [comment]
marked this BP as `accepted` with 3 +1 and 0 -1.

**Pull request merged**